### PR TITLE
Move impala ddl to base_sql

### PR DESCRIPTION
**Pull request by matthewmturner**
@datapythonista i started with what was in the spark ddl file.  still have some work to do checking tests and other backends but this gets the ball rolling.  hopefully CI is ok. will keep you posted.

### CODE REVIEW

**Comment by matthewmturner** [comment]
@datapythonista made a bunch of updates to tests and some other parts of spark.  went with similar approach to what we did with impala compiler to try reducing duplications and using just_base_sql classes where possible(i.e. if just subclassing from base with no modifications then dont bother creating within other backends and just import base classes as needed).  hopefully this is what you were envisioning....will see how CI turns out.

let me know if any thoughts!

**Comment by datapythonista** [comment]
Thanks @matthewmturner looks great. Seems like there is a bug, but other than that looks good to me.
```
Traceback (most recent call last):
  File "/ibis/ci/impalamgr.py", line 352, in <module>
    main()
  File "/opt/conda/envs/ibis-env/lib/python3.8/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/opt/conda/envs/ibis-env/lib/python3.8/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/opt/conda/envs/ibis-env/lib/python3.8/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/opt/conda/envs/ibis-env/lib/python3.8/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/opt/conda/envs/ibis-env/lib/python3.8/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/ibis/ci/impalamgr.py", line 260, in load
    con = make_ibis_client(ENV)
  File "/ibis/ci/impalamgr.py", line 49, in make_ibis_client
    return ibis.impala.connect(
  File "/ibis/ibis/impala/api.py", line 122, in connect
    client = ImpalaClient(con, hdfs_client=hdfs_client)
  File "/ibis/ibis/impala/client.py", line 789, in __init__
    self._ensure_temp_db_exists()
  File "/ibis/ibis/impala/client.py", line 1399, in _ensure_temp_db_exists
    self.create_database(name, path=path, force=True)
  File "/ibis/ibis/impala/client.py", line 925, in create_database
    statement = ddl.CreateDatabase(name, path=path, can_exist=force)
AttributeError: module 'ibis.impala.ddl' has no attribute 'CreateDatabase'
```

**Comment by matthewmturner** [comment]
@datapythonista i think its going to be good now after running isort.  i didnt add everything from the `impala.ddl` as i wasnt sure how much truly belong in `base_sql` or if it was impala specific.  can you check out whats remaining and let me know if you want anything else moved?

**Comment by matthewmturner** [comment]
> 
> 
> @datapythonista i think its going to be good now after running isort. i didnt add everything from the `impala.ddl` as i wasnt sure how much truly belong in `base_sql` or if it was impala specific. can you check out whats remaining and let me know if you want anything else moved?

spoke too soon...had to make some more updates.  but the ask about checking out the impala ddl and seeing if other classes should move still applies.

Thanks!

**Comment by matthewmturner** [comment]
@datapythonista all green. let me know your thoughts!

**Review by datapythonista — APPROVED** [review]
Thanks @matthewmturner, looks great.

If other backends don't need to import impala, then all the generic stuff should be moved. I'm sure there will be space for more refactoring later, also considering merging `base_sql` and `base_sqlalchemy` and other things. But for now this is extremely useful. It should bring stability to the CI, and simplify things a lot.

**Comment by matthewmturner** [comment]
for sure makes sense.  my issue was that the only way i currently had to tell if something was impala specific or generic was if the class/function had impala in the name or there was comment referencing impala specifics.  i wasnt sure if i could assume everything else was generic.  unfortunately i dont have enough experience with different sql implementations to know whats specific to one vs being generic to all.

File: ibis/impala/ddl.py [hunk 1]
```diff
@@ -761,38 +313,14 @@ def compile(self):
...
-        self, name, inputs, must_exist=True, aggregate=False, database=None
-    ):
-        super().__init__(must_exist=must_exist)
-        self.name = name
-        self.inputs = tuple(map(dt.dtype, inputs))
-        self.must_exist = must_exist
-        self.aggregate = aggregate
-        self.database = database
-
+class ImpalaDropFunction(DropFunction):
```
Comment (jreback) [inline ibis/impala/ddl.py:316]: is there a reason you renamed this? shouldn't it just be DropFunction? (and change the import to avoid the name conflict)

**Review by jreback — CHANGES_REQUESTED** [review]

File: ibis/impala/ddl.py (hunk 1 above)
Comment (matthewmturner) [reply ibis/impala/ddl.py:316]: there was a function with impala in the name so i figured that was a specific implementation for impala.  so i took that part out for the `base_sql` implementation and made this `ImpalaDropFunction` with that function added.

do you think that this function isnt specific to impala and can just be moved to `DropFunction` in `base_sql`?

Comment (jreback) [reply ibis/impala/ddl.py:316]: i think the location is fine, just puzzled why you added the Impala prefix; we don't do this anywhere else; i would just rename it to DropFunction and call it a day.

Comment (matthewmturner) [reply ibis/impala/ddl.py]: Just updated and CI is green. Let me know if anything else.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thank @matthewmturner very nice !

**Pull request merged**